staging/src/k8s.io/client-go/tools/cache/controller_test.go

```diff
@@ -572,5 +579,173 @@ func TestTransformingInformer(t *testing.T) {
 	verifyEvent(watch.Deleted, expectedPod("pod1", "2"), nil)
 	verifyStore([]interface{}{expectedPod("pod2", "2"), expectedPod("pod3", "1")})
 
+	// Delete pod and ensure the watch event never hit the client. Then trigger a relist,
+	// that should cause the controller to pick up the deletion and propagate it as an DeletedFinalStateUnknown object.
+	source.DeleteDropWatch(makePod("pod2", "3"))
+	source.ResetWatch()
+
+	verifyEvent(watch.Modified, expectedPod("pod3", "1"), expectedPod("pod3", "1"))
+	verifyEvent(watch.Deleted, expectedPod("pod2", "2"), nil)
+
+	verifyStore([]interface{}{expectedPod("pod3", "1")})
+
+	// make controller block on {Deleted, "pod3"}, so that {Added, "pod4"} will still be in the queue
+	// when the watch is terminated and a relist occurs, and "pod4" is missing.
+	source.Delete(makePod("pod3", "1"))
+	source.Add(makePod("pod4", "1"))
+	source.Modify(makePod("pod4", "2"))
+	source.DeleteDropWatch(makePod("pod4", "3"))
+	source.ResetWatch()
+
+	verifyEvent(watch.Deleted, expectedPod("pod3", "1"), nil)
+	verifyEvent(watch.Added, nil, expectedPod("pod4", "1"))
+	verifyEvent(watch.Modified, expectedPod("pod4", "1"), expectedPod("pod4", "2"))
+	verifyEvent(watch.Deleted, expectedPod("pod4", "2"), nil)
+
+	verifyStore([]interface{}{})
+}
+
+func TestTransformingInformerWithCustomObjectType(t *testing.T) {
+	// source simulates an apiserver object endpoint.
+	source := fcache.NewFakeControllerSource()
+
+	makePod := func(name, generation string) *v1.Pod {
+		return &v1.Pod{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: "namespace",
+				Labels:    map[string]string{"generation": generation},
+			},
+		}
+	}
+	expectedObjectMeta := func(name, generation string) *metav1.ObjectMeta {
+		meta := makePod(name, generation).ObjectMeta
+		return &meta
+	}
+
+	source.Add(makePod("pod1", "1"))
+	source.Modify(makePod("pod1", "2"))
+
+	type event struct {
+		eventType watch.EventType
+		previous  interface{}
+		current   interface{}
+	}
+	events := make(chan event, 1)
+	recordEvent := func(eventType watch.EventType, previous, current interface{}) {
+		events <- event{eventType: eventType, previous: previous, current: current}
+	}
+	verifyEvent := func(eventType watch.EventType, previous, current interface{}) {
```

Comment:
This is a lot of code to copy from the test above, maybe consider inserting a commit which refactors to make it reusable? (preferably keeping refactoring code and test changing code in separate commits)